pkg/image/api/types.go

```diff
@@ -60,6 +77,41 @@ type ImageLayer struct {
 	Size int64
 }
 
+// SignatureValidationResult holds a result of signature validation.
+type SignatureValidationResult string
+
+// ImageSignature holds a signature of an image.
+type ImageSignature struct {
+	// Describes a type of stored blob. Provided by caller.
+	Type string
+	// An opaque binary string which is an image's signature.
+	Content []byte
+	// A time of last server-side validation.
+	ValidatedOn unversioned.Time
+	// Result of validation performed at ValidatedOn.
+	ValidationResult SignatureValidationResult
+
+	// At most one of the following pointers can point to details. If the signature type isn't recognized,
+	// they will all be nil.
+
+	// Details for signature of type "atomic".
+	AtomicSignatureDetails *AtomicSignatureDetails
+	// For any future supported signature type, add additional pointer to its own details struct.
+}
+
+// AtomicSignatureDetails holds metadata of recognized signature of type "atomic". They are parsed from
+// signature's content.
+type AtomicSignatureDetails struct {
+	// Digest of signed manifest.
+	SignedImageName string
+	// Docker image reference
+	SignedDockerImageReference string
```

Comment:
> Define more strongly what this is.

This is the “identity” of the image, what the signature claims the image to be:

- It is the difference between knowing that an image is signed by Red Hat (but is a 10-year-old RHEL 5 with hypothetical architecturally unfixable and well-known vulnerabilities) and that the image is signed by Red Hat, who claims that this is their RHEL 7.5.10.
- Actually it is a somewhat of a starting point for signature verification: I need to know that the signature claims this to be a Red Hat-created image _before_ I can decide that the signature should be validated against a known Red Hat public key¹.

Note that the claimed identity does not change by copying: A RHEL7.5.10 image will keep the signature claiming the identity `registry.access.redhat.com/rhel7/rhel:7.5.10` (or whatever the format is) even if it is copied across 5 different registries using 3 different protocols. (If an ISV copies and republishes an image as `isv.com/my/superos`, they can add a signature claiming this identity with their public key, and they can decide to keep or drop Red Hat’s `rhel7/rhel:7.5.10` signature.)

> I would not expect us to be validating against this in the future, tbh, because we want images to have many different references.

What does this mean? Having the same image referenced by several imagestreams, possibly in different namespaces? If so, the “claimed identity” does not change.


¹ Actually we would probably do the signature verification first, and key matching later—but that is an implementation detail. Matching public keys and identities is essential, otherwise anyone trusted can spoof images supposedly published by anyone else.